### ISSUE TITLE
1-5 to 1-7 upgrade test: Deployment overlapping deployment should not fight with each other

### ISSUE THREAD

**Issue by monopole** [kubernetes/kubernetes#47766]
[gubernator](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gke-cvm-1-5-cvm-1-7-upgrade-cluster/18#k8sio-deployment-overlapping-deployment-should-not-fight-with-each-other)
[history](https://k8s-testgrid.appspot.com/1.5-1.7-upgrade#gke-cvm-1-5-cvm-1-7-upgrade-cluster&include-filter-by-regex=%5EOverall%24%7C%5C%5Bk8s%5C.io%5C%5D%5C%20Deployment%5C%20overlapping%5C%20deployment%5C%20should%5C%20not%5C%20fight%5C%20with%5C%20each%5C%20other)

Failing as of 6/16

```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:98
Failed to update the second deployment's overlapping annotation
Expected error:
    <*errors.errorString | 0xc4203608c0>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
not to have occurred
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:1223
```

/sig apps
/assign @enisoc

**Comment by enisoc** [comment]
It seems like this is running the 1.5 e2e binary against a 1.7 cluster. The annotation was no longer used as of 1.6, so the 1.5 version of this test will not work on 1.6+.

**Comment by janetkuo** [comment]
We can either remove this upgrade test, or modify 1.5 e2e test to skip server version 1.6+

**Comment by enisoc** [comment]
Is that something we do in general? Backport fixes to old e2e tests to make them work against newer clusters whose expected behavior has changed?

**Comment by janetkuo** [comment]
> Backport fixes to old e2e tests to make them work against newer clusters whose expected behavior has changed?

Needing to do that means we either wrote a bad e2e test or broke backward compatibility.

**Comment by enisoc** [comment]
> Needing to do that usually means we either wrote a bad e2e test or broke backward compatibility.

In retrospect, I think the annotation was more of an implementation detail (which is why we felt justified removing it in 1.6), so the e2e should have tested some broader definition of "protection against overlapping selectors".

However, at this point in the release cycle I think it would take too long to refactor the test, so I'd be ok with skipping it for 1.6+. Is there already a function that can detect the server version in an e2e test?

**Comment by janetkuo** [comment]
> In retrospect, I think the annotation was more of an implementation detail (which is why we felt justified removing it in 1.6), so the e2e should have tested some broader definition of "protection against overlapping selectors".

Agree. We should document this in the community "how to write e2e tests" guide too

>  Is there already a function that can detect the server version in an e2e test?

Yes there is. I'll file a PR

**Comment by janetkuo** [comment]
This is fixed by #47886
